python/pyspark/rdd.py

```diff
@@ -121,15 +121,30 @@ def _parse_memory(s):
 
 
 def _load_from_socket(port, serializer):
-    sock = socket.socket()
-    sock.settimeout(3)
-    try:
-        sock.connect(("localhost", port))
-        rf = sock.makefile("rb", 65536)
-        for item in serializer.load_stream(rf):
-            yield item
-    finally:
-        sock.close()
+    sock = None
+    # Support for both IPv4 and IPv6.
+    # On most of IPv6-ready systems, IPv6 will take precedence.
+    for res in socket.getaddrinfo("localhost", port, socket.AF_UNSPEC, socket.SOCK_STREAM):
+        af, socktype, proto, canonname, sa = res
+        try:
+            sock = socket.socket(af, socktype, proto)
+            sock.settimeout(3)
+            sock.connect(sa)
+        except socket.error:
+            sock = None
+            continue
+        break
+    if sock:
+        try:
+            rf = sock.makefile("rb", 65536)
+            for item in serializer.load_stream(rf):
+                yield item
+        except socket.error:
+            raise Exception("encounter error when connecting to socket server")
+        finally:
+            sock.close()
+    else:
+        raise Exception("could not open socket")
```

Comment:
Hi Davies, you mean the "else branch"? I think it will throw the exception once connections in both IPV4 and IPV6 cannot be established. What's the benefit to move the "else" branch?